Enforce polymorphic types are only instantiated with nodes

NathanReb opened the pull request:
Depends on #68 so I'm marking as WIP until it's merged!

This modifies `Grammar` so that polymorphic types from a grammar can only be instantiated by another named type it declares.

NathanReb committed:
Enforce polymoprhic types are only instantiated with nodes in Grammar

NathanReb commented:
Should be ready for review!

ceastlund commented on astlib/grammar.ml:
```diff
@@ -10,7 +10,7 @@
     [Name "expression"] represents the type [expression].
 
-    [Instance ("class_infos", [Name "class_expr"])] represents the type
+    [Instance ("class_infos", [Targ_name "class_expr"])] represents the type
```
Should be `Tname`, not `Targ_name`.

ceastlund commented on astlib/unstable_for_testing.ml:
```diff
@@ -14,7 +14,7 @@ let rec update_ty ty : Grammar.ty =
   | Option ty -> Option (update_ty ty)
   | Tuple tuple -> Tuple (update_tuple tuple)
-  | Instance (name, tuple) -> Instance (name, update_tuple tuple)
+  | Instance _ -> ty
```
This should be applying a `List.rev` to the args.

ceastlund commented on ast/cinaps/grammar.mli:
```diff
@@ -1,5 +1,9 @@
-val string_of_ty : internal: bool -> Astlib.Grammar.ty -> string
+val string_of_targ : ?nodify: unit -> internal: bool -> Astlib.Grammar.targ -> string
+
+val string_of_ty : ?nodify: unit -> internal: bool -> Astlib.Grammar.ty -> string
```
Let's document what `?nodify` does and why we need it. It was not obvious on my first read, I had to go digging around.

ceastlund commented on ast/cinaps/grammar.mli:
```diff
@@ -1,5 +1,9 @@
+val string_of_ty : ?nodify: unit -> internal: bool -> Astlib.Grammar.ty -> string
+
 val string_of_tuple_type :
+  ?nodify: unit ->
```
I'm not a fan of the optional `unit` argument convention. We don't "cut" the default value out of other optional types, e.g. with an optional `int` and a default of 0 we don't document it as an optional "non-zero integer". I'd prefer `?nodify:bool` and a documented default. I also think `~nodify:true` and `nodify:false` read better than `~nodify:()` and `?nodify:None`.

NathanReb replied:
Moving it back to a boolean sounds good to me!

ceastlund commented on ast/cinaps/grammar.ml:
```diff
@@ -1,24 +1,37 @@
+  | None -> Ml.tvar var
+  | Some () -> Ml.poly_inst "node" ~args:[Ml.tvar var]
+
+let string_of_targ ?nodify ~internal targ =
```
For purely internal functions `string_of_var` and `string_of_targ`, if we make `nodify` a required argument, we can never forget to pass it in.

ceastlund commented on ast/cinaps/name.ml:
```diff
@@ -31,8 +19,8 @@ let fix_builtin name =
     (String.concat ~sep:"_"
        (List.map ~f:Ml.id
-          (prefix @ add_tuple_suffix tys [])))
+          (prefix @ List.fold_right targs ~init:[] ~f:add_targ_suffix)))
```
At this point, it could just be `List.map targs ~f:name_of_targ`, and we could leave out the `fold_right`.

ceastlund reviewed (changes requested):
Left some comments. Biggest question mark is the `?nodify` optional argument. I think I get it, but it's an odd sort of operation.

Do we have any places where we have uninstantiated type variables, but we _don't_ want to nodify them?

NathanReb commented:
I addressed your comments!

Regarding the `nodify` argument I agree it's a bit odd but `Gen_version` relied on `Grammar.ty` not having this new constraint on type parameters to "rewrite" variables as instances of `node`. Now that grammar can't be used to represent such types we'd either have to use a different representation in `Gen_version` or to just print variables differently when we need to.

> Do we have any places where we have uninstantiated type variables, but we don't want to nodify them?

`Gen_version` is the only place where we do it atm and even in that module we want to be able to print variables normally. It seems to work fine that way but it might make sense to nodify variables in some other places to further constraint the API.

I think the main issue here is that we use `Grammar` as an intermediate representation for OCaml types in our code generators when it wasn't specifically designed for that purpose. To be clearer it kind of was but the set of constraints we want for `Grammar` isn't necessarily relevant to our code generation logic. Maybe we need another set of types that we use in the cinaps code and add `to/from_grammar` converters.

Let me know what you think!

ceastlund reviewed (approved):
Looks good!
